Add graph-tool

stuarteberg opened the pull request:
This PR adds a recipe for [`graph-tool`][graph-tool], a Python library for graph analysis based on [`boost-graph`][boost-graph].

This is an unusual recipe in a few respects:

1. On mac, this package requires clang-8.0, rather than conda's default (clang-4.0).
2. The build requires a lot of RAM, and may fail intermittently ("internal compiler error") as a result.  Therefore, `build.sh` simply restarts `make` a few times in the hope that it will eventually succeed.
3. The build will take a long time (hours).  I think the Azure provider can handle it without timing out, but the others probably can't.  To avoid tying up several build queues for multiple hours every time this package needs to be built, I'm skipping all configurations except for python 3.7.

Other notes:

- Although `graph-tool` has plotting capabilities, I'm not attempting to support those yet.  (Still missing some dependencies.)



[graph-tool]: https://graph-tool.skewed.de
[boost-graph]: http://www.boost.org/doc/libs/release/libs/graph

<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with
a PR or once it's ready for review, please let the right people know.
There are language-specific teams for reviewing recipes.

Currently available teams are:
- python `@conda-forge/help-python`
- python/c hybrid `@conda-forge/help-python-c`
- r `@conda-forge/help-r`
- java `@conda-forge/help-java`
- nodejs `@conda-forge/help-nodejs`
- c/c++ `@conda-forge/help-c-cpp`
- perl `@conda-forge/help-perl`
- Julia `@conda-forge/help-julia`

If your PR doesn't fall into those categories please contact
the full review team `@conda-forge/staged-recipes`.

Due to GitHub limitations first time contributors to conda-forge are unable
to ping these teams.
Consider asking on our [Gitter channel](https://gitter.im/conda-forge/conda-forge.github.io)
if your recipe isn't reviewed promptly.
-->

---

Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vend other packages
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your
      recipe (see [here][1] for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there **[I confirm]**

[1]: https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/graph-tool```) and found some lint.

Here's what I've got...


For **recipes/graph-tool**:

 * Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form. See lines [25, 42]

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/graph-tool```) and found it was in an excellent condition.

stuarteberg commented:
@conda-forge/help-python-c

This recipe is ready for review.  (The CI is re-running because I pushed again, but it should pass again. BTW, how do I cancel a build?)

There are two sticking points for which I need advice:

1. Is it okay to use clang-8.0?  (This recipe requires it.)
2. This recipe needs 5+ hours to build.  I increased the Azure timeout.  Is that okay?  If you'd prefer me not to commit that change here in `staged-recipes`, I can revert it and then re-apply it only in the `graph-tool-feedstock` (once it's created).

Frankly, I'm just pleased that this package builds at all...

xhochy commented on recipes/example/meta.yaml:
```diff
@@ -1,82 +0,0 @@
-# Note: there are many handy hints in comments in this example -- remove them when you've finalized your recipe
```
Please don't delete the example

stuarteberg replied:
@xhochy -- Sorry about that.  I've restored it now.

scopatz commented on .azure-pipelines/azure-pipelines-linux.yml:
```diff
@@ -9,7 +9,7 @@ jobs:
         CF_MAX_PY_VER: 37
         AZURE: True
-  timeoutInMinutes: 240
+  timeoutInMinutes: 360
```
Please don't modify files outside of the recipes/ dir in a recipe PR

scopatz commented on .azure-pipelines/azure-pipelines-osx.yml:
```diff
@@ -8,7 +8,7 @@ jobs:
         CONFIG: azure-osx-64-comp7
         CF_MAX_PY_VER: 37
-  timeoutInMinutes: 240
+  timeoutInMinutes: 360
```
Same

stuarteberg commented:
@scopatz Thanks for the review.  I've reverted the non-recipe changes and squashed all commits.  I think this is ready for merge, unless you have any other comments.

(The build will now fail due to the reduced timeout, but it will pass again once it becomes a feedstock.  Feedstocks have higher timeouts.)

stuarteberg commented:
@conda-forge/help-python -- This is ready for merge.  Thanks for the reviews, @xhochy and @scopatz!

As noted above, you can ignore the CI timeout issue.  With the normal feedstock timeout (6 hours), this build [passes][].

[passes]: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=65154

isuruf commented on recipes/graph-tool/meta.yaml:
```diff
@@ -0,0 +1,80 @@
+about:
+  home: https://graph-tool.skewed.de/
+  license: GPL-3.0
+  license_file: LICENSE
```
You need to package the license of pcg-cpp

stuarteberg replied:
Whoops, thanks.  Done.

isuruf commented:
@stuarteberg, can you send a separate PR to increase the timeout to 360 minutes

isuruf commented on recipes/graph-tool/build.sh:
```diff
@@ -0,0 +1,80 @@
+    export LDFLAGS="${LDFLAGS} -undefined dynamic_lookup"
+
+    # https://github.com/conda-forge/conda-forge-ci-setup-feedstock/issues/53#issuecomment-490314561
+    export CPU_COUNT=4
```
Please do this only if `"$CI" == "azure"`.

stuarteberg replied:
done

isuruf commented on recipes/graph-tool/build.sh:
```diff
@@ -0,0 +1,80 @@
+# This is a suitable no-op.
+PYTHON_LIBS=" "
+
+if [[ $(uname) == Darwin ]]; then
```
Use `$target_platform == osx*` instead of `uname`

stuarteberg replied:
done

stuarteberg commented:
@isuruf

>can you send a separate PR to increase the timeout to 360 minutes

Done! See #9244

isuruf commented:
Can you rebase on top of master?

stuarteberg commented:
>Can you rebase on top of master?

Done.

isuruf commented:
Please fix the following warnings,
```
WARNING (graph-tool,lib/python3.7/site-packages/graph_tool/generation/libgraph_tool_generation.so): Needed DSO lib/libgmp.so.10 found in ['gmp']
WARNING (graph-tool,lib/python3.7/site-packages/graph_tool/generation/libgraph_tool_generation.so): .. but ['gmp'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
WARNING (graph-tool,lib/python3.7/site-packages/graph_tool/draw/libgraph_tool_draw.so): Needed DSO lib/libboost_context.so.1.70.0 found in ['boost-cpp']
WARNING (graph-tool,lib/python3.7/site-packages/graph_tool/draw/libgraph_tool_draw.so): .. but ['boost-cpp'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
```

stuarteberg commented:
@isuruf 

>Please fix the following warnings

Thanks for catching those.  Now fixed.

stuarteberg commented:
Argh.  I missed one on Mac:

```
WARNING (graph-tool,lib/python3.7/site-packages/graph_tool/stats/libgraph_tool_stats.0.so): Needed DSO lib/libomp.dylib found in ['llvm-openmp']
WARNING (graph-tool,lib/python3.7/site-packages/graph_tool/stats/libgraph_tool_stats.0.so): .. but ['llvm-openmp'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
```

Fixing now.
